File: 在排序数组中查找元素的第一个和最后一个位置.py
class Solution:
    def lower_bound(self, nums: List[int], target: int) -> List[int]:
        n = len(nums)
        left, right = 0, n - 1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        return left

    def searchRange(self, nums: List[int], target: int) -> List[int]:
        n = len(nums)
        if n == 0:
            return [-1, -1]
        l = self.lower_bound(nums, target)
        if l > n - 1 or nums[l] != target:
            return [-1, -1]
        r = self.lower_bound(nums, target + 1) - 1
        return [l, r]

File: 在排序数组中查找元素的第一个和最后一个位置.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int left = ranges::lower_bound(nums, target) - nums.begin();
        if (left == nums.size() or nums[left] ^ target) {
            return {-1, -1};
        }
        int right = ranges::upper_bound(nums, target) - nums.begin() - 1;
        return {left, right};
    }
};
